File: src/org/dyndns/fzoli/rccar/model/host/HostData.java
package org.dyndns.fzoli.rccar.model.host;

import java.io.Serializable;
import java.util.Date;
import org.dyndns.fzoli.rccar.model.BaseData;
import org.dyndns.fzoli.rccar.model.BatteryPartialBaseData;
import org.dyndns.fzoli.rccar.model.Control;
import org.dyndns.fzoli.rccar.model.ControlPartialBaseData;
import org.dyndns.fzoli.rccar.model.PartialBaseData;
import org.dyndns.fzoli.rccar.model.Point3D;

/**
 * Egy hosztra jellemző összes adat, ami a telefon szenzoraiból jön.
 * Amikor a telefon a szenzoroktól új adatot kap, megnézi, hogy a 'refresh time'
 * szerint kell-e üzenni a hídnak.
 * Ha kell üzenni, megnézi, hogy egy vagy több adat változott-e.
 * Ha több adat változott, a hídnak teljes modelt, egyébként részmodelt küld.
 * A híd a saját modeljét frissíti a kapott adat alapján az update metódussal.
 * Kivételt képez az akkumulátor-szint változás - mivel az ritka esemény -, mely
 * változása mindig részadatban érkezik és nem függ a 'refresh time'-tól.
 * @author zoli
 */
public class HostData extends BaseData<HostData, PartialBaseData<HostData, ?>> {
    
    /**
     * A HostData részadata.
     * Egy HostPartialData objektumot átadva a HostData objektumnak, egyszerű frissítést lehet végrehajtani.
     */
    public static abstract class PartialHostData<T extends Serializable> extends PartialBaseData<HostData, T> {

        /**
         * Részadat inicializálása és beállítása.
         * @param data az adat
         */
        protected PartialHostData(T data) {
            super(data);
        }
        
    }
    
    /**
     * A HostData vezérlő részadata, ami egy boolean érték változását tartalmazza.
     */
    public static class BooleanPartialHostData extends PartialHostData<Boolean> {

        /**
         * A HostData Boolean változóinak megfeleltetett felsorolás.
         */
        public static enum BooleanType {
            STREAMING,
            VEHICLE_CONNECTED,
            UP_2_DATE
        }
        
        /**
         * Megmondja, melyik adatról van szó.
         */
        private final BooleanType type;
        
        /**
         * Részadat inicializálása és beállítása.
         * @param data a vezérlőjel
         * @param type melyik változó
         */
        public BooleanPartialHostData(Boolean data, BooleanType type) {
            super(data);
            this.type = type;
        }

        /**
         * Alkalmazza az új értéket a paraméterben megadott adaton.
         */
        @Override
        public void apply(HostData d) {
            if (d != null && type != null) {
                switch (type) {
                    case STREAMING:
                        d.setStreaming(data);
                        break;
                    case VEHICLE_CONNECTED:
                        d.setVehicleConnected(data);
                        break;
                    case UP_2_DATE:
                        d.setUp2Date(data);
                }
            }
        }
        
    }
    
    /**
     * A HostData részadata, ami egy vagy több pont változását tartalmazza.
     */
    public static class PointPartialHostData extends PartialHostData<PointPartialHostData.PointData[]> {
        
        /**
         * A HostData Point3D változóinak megfeleltetett felsorolás.
         */
        public static enum PointType {
            GPS_POSITION,
            GRAVITATIONAL_FIELD,
            MAGNETIC_FIELD
        }
        
        /**
         * Egy 3D pont három koordinátáját és típusát tartalmazó bab.
         */
        public static class PointData implements Serializable {
            
            /**
             * A 3D pont.
             */
            private final Point3D point;
            
            /**
             * A 3D pont típusa.
             */
            private final PointType type;

            /**
             * Konstruktor.
             */
            public PointData(Point3D point, PointType type) {
                this.point = point;
                this.type = type;
            }
            
        }

        /**
         * Részadat inicializálása és beállítása.
         * Ezzel a verzióval 847 bájt egy objektum mérete szerializálva.
         * Ha csak egy 3D pont lenne tárolva, 611 bájt helyigénye lenne,
         * de két pont esetén már 1222 bájt helyre lenne szükség, ellenben
         * az aktuális megoldással csak 888 bájtot igényel két koordináta.
         * Mivel nagyon ritka az az eset, hogy csak egy szenzoradat módosul,
         * ezért ez a megoldás kevesebb adatforgalmat generál, mint az előző.
         * @param data a 3D pontadatok
         */
        public PointPartialHostData(PointData ... data) {
            super(data);
        }
        
        /**
         * Alkalmazza a 3D pontot/pontokat a paraméterben megadott adaton.
         * @param d a teljes adat, amin a módosítást alkalmazni kell
         */
        @Override
        public void apply(HostData d) {
            if (d != null && data != null && data.length > 0) {
                d.pointChanging = true;
                for (int i = 0; i < data.length; i++) {
                    if (i == data.length - 1) d.pointChanging = false;
                    PointData pd = data[i];
                    if (pd.type != null) switch (pd.type) {
                        case GPS_POSITION:
                            d.setGpsPosition(pd.point);
                            break;
                        case GRAVITATIONAL_FIELD:
                            d.setGravitationalField(pd.point);
                            break;
                        case MAGNETIC_FIELD:
                            d.setMagneticField(pd.point);
                    }
                }
            }
        }
        
    }
    
    /**
     * A HostData részadata, ami az akkumulátorszint változását tartalmazza.
     */
    public static class BatteryPartialHostData extends BatteryPartialBaseData<HostData> {

        /**
         * Részadat inicializálása és beállítása.
         * @param data az akkumulátorszint
         */
        public BatteryPartialHostData(Integer data) {
            super(data);
        }
        
    }
    
    /**
     * A HostData vezérlő részadata, ami az autó irányításában játszik szerepet.
     */
    public static class ControlPartialHostData extends ControlPartialBaseData<HostData> {

        /**
         * Részadat inicializálása és beállítása.
         * @param data a vezérlőjel
         */
        public ControlPartialHostData(Control data) {
            super(data);
        }
        
    }
    
    /**
     * GPS koordináta.
     */
    private Point3D gpsPosition;
    
    /**
     * Gravitációs mező.
     */
    private Point3D gravitationalField;
    
    /**
     * Mágneses mező.
     */
    private Point3D magneticField;
    
    /**
     * Az utolsó módosulás előtti szenzor-adat.
     */
    private Point3D previousGpsPosition, previousMagneticField, previousGravitationalField;
    
    /**
     * Az utolsó előtti GPS pozíció módosulásának dátuma.
     */
    private Date previousGpsChangeDate;
    
    /**
     * Az utolsó GPS pozíció módosulásának dátuma.
     */
    private Date gpsChangeDate;
    
    /**
     * Folyamatban van-e az MJPEG streamelés.
     * Kezdetben nincs streamelés.
     */
    private Boolean streaming = false;
    
    /**
     * Megmondja, hogy a telefonhoz csatlakoztatva van-e a jármű.
     */
    private Boolean vehicleConnected = false;
    
    /**
     * Megmondja, hogy folyamatban van-e a Point3D adatok beállítása.
     */
    private boolean pointChanging = false;
    
    /**
     * Az északtól való eltéréshez hozzáadódó érték.
     */
    private Integer additionalDegree;
    
    /**
     * Konstruktor Híd oldalra.
     * Kezdetben nem ismert minden adat,
     * az adat frissítésére a hoszt kapcsolódásakor kerül sor.
     */
    public HostData() {
        super();
        init();
    }
    
    /**
     * Konstruktor hoszt oldalra.
     * Az állandó értékek konstruktorból állítódnak be.
     */
    public HostData(boolean fullX, boolean fullY, int additionalDegree) {
        super();
        setFullX(fullX);
        setFullY(fullY);
        setAdditionalDegree(additionalDegree);
        init();
    }
    
    /**
     * Beállítja a kezdeti paramétereket.
     * Kezdetben a jármű áll.
     */
    private void init() {
        setControl(new Control(0, 0));
    }
    
    /**
     * Megadja, hogy folyamatban van-e a streamelés.
     */
    public Boolean isStreaming() {
        return streaming;
    }

    /**
     * Megmondja, hogy a telefonhoz csatlakoztatva van-e a jármű.
     */
    public Boolean isVehicleConnected() {
        return vehicleConnected;
    }

    /**
     * Megadja, hogy a GPS adat naprakész-e.
     * Ha nincs megadva, akkor hamis értékkel tér vissza.
     */
    @Override
    public Boolean isUp2Date() {
        Boolean b = super.isUp2Date();
        return b == null ? false : b;
    }

    /**
     * Megmondja, hogy folyamatban van-e a Point3D adatok beállítása.
     * A szerver oldalán hasznos metódus, mert segítségével elkerülhető
     * a fölösleges adatfeldolgozás, amikoris több Point3D kerül kiküldésre.
     * @return true, ha éppen valamelyik Point3D adat állítódik be és miután beállítódott,
     * még legalább egy Point3D adat beállítás meg fog történni; egyébként false.
     */
    protected boolean isPointChanging() {
        return pointChanging;
    }

    /**
     * Megadja az északtól való eltéréshez hozzáadódó értéket.
     */
    public Integer getAdditionalDegree() {
        return additionalDegree;
    }
    
    /**
     * Megadja a GPS koordinátát.
     */
    public Point3D getGpsPosition() {
        return gpsPosition;
    }
    
    /**
     * Megadja a gravitációs mező erősségét.
     */
    public Point3D getGravitationalField() {
        return gravitationalField;
    }

    /**
     * Megadja a mágneses mező erősségét.
     */
    public Point3D getMagneticField() {
        return magneticField;
    }

    /**
     * Megadja az utolsó módosítás előtti GPS pozíciót.
     */
    public Point3D getPreviousGpsPosition() {
        return previousGpsPosition;
    }

    /**
     * Megadja az utolsó módosítás előtti gravitációs mező erősségét.
     */
    public Point3D getPreviousGravitationalField() {
        return previousGravitationalField;
    }
    
    /**
     * Megadja az utolsó módosítás előtti mágneses mező erősségét.
     */
    public Point3D getPreviousMagneticField() {
        return previousMagneticField;
    }
    
    /**
     * Megadja az utolsó GPS pozíció módosulásának dátumát.
     */
    public Date getGpsChangeDate() {
        return gpsChangeDate;
    }

    /**
     * Megadja az utolsó előtti GPS pozíció módosulásának dátumát.
     */
    public Date getPreviousGpsChangeDate() {
        return previousGpsChangeDate;
    }

    /**
     * Beállítja, van-e streamelés.
     * @param streaming ha null, (az alapértelmezett) false állítódik be
     */
    public void setStreaming(Boolean streaming) {
        if (streaming == null) streaming = false;
        this.streaming = streaming;
    }

    /**
     * Beállítja, hogy a telefonhoz csatlakoztatva van-e a jármű.
     */
    public void setVehicleConnected(Boolean vehicleConnected) {
        this.vehicleConnected = vehicleConnected;
    }
    
    /**
     * Beállítja az autó vezérlőjelét.
     * @param controll ha null, (az alapértelmezett) 0;0 állítódik be
     */
    @Override
    public void setControl(Control controll) {
        if (controll == null) controll = new Control(0, 0);
        super.setControl(controll);
    }

    /**
     * Beállítja az északtól való eltéréshez hozzáadódó értéket.
     */
    public void setAdditionalDegree(Integer additionalDegree) {
        this.additionalDegree = additionalDegree;
    }
    
    /**
     * Beállítja a GPS koordinátát.
     * Még mielőtt megváltozna az adat, az előző adat eltárolódik.
     */
    public void setGpsPosition(Point3D gpsPosition) {
        previousGpsChangeDate = gpsChangeDate;
        gpsChangeDate = new Date();
        this.previousGpsPosition = getGpsPosition();
        this.gpsPosition = gpsPosition;
    }

    /**
     * Beállítja a gravitációs mező erősségét.
     */
    public void setGravitationalField(Point3D gravitationalField) {
        this.previousGravitationalField = getGravitationalField();
        this.gravitationalField = gravitationalField;
    }

    /**
     * Beállítja a mágneses mező erősségét.
     */
    public void setMagneticField(Point3D magneticField) {
        this.previousMagneticField = getMagneticField();
        this.magneticField = magneticField;
    }
    
    /**
     * Frissíti az adatokat a megadott adatokra.
     * @param d az új adatok
     */
    @Override
    public void update(HostData d) {
        if (d != null) {
            setStreaming(d.isStreaming());
            setVehicleConnected(d.isVehicleConnected());
            setGpsPosition(d.getGpsPosition());
            setGravitationalField(d.getGravitationalField());
            setMagneticField(d.getMagneticField());
            setAdditionalDegree(d.getAdditionalDegree());
            super.update(d);
        }
    }

    /**
     * Kinullázza azokat a szenzoradatokat, melyek azonnal megtudhatóak, tehát nem igényelnek online kapcsolatot (minden, ami nem GPS).
     * A többi adatot megtartja, hátha elvész a kapcsolat és csak az utolsó adat lesz elérhető.
     */
    @Override
    public void clear() {
        super.clear();
        previousMagneticField = magneticField = null;
        previousGravitationalField = gravitationalField = null;
    }
    
}

File: cpp/RCServer/HostData.cpp
/* 
 * File:   HostData.cpp
 * Author: zoli
 * 
 * Created on 2013. szeptember 18., 1:37
 */

#include "HostData.h"
#include "BatteryPartialHostData.h"
#include "BooleanPartialHostData.h"
#include "ControlPartialHostData.h"
#include "PointPartialHostData.h"
#include "SpeedPartialHostData.h"

REGISTER_DEF_TYPE(HostData, org.dyndns.fzoli.rccar.model.host.HostData);

HostData::HostData() {
    streaming = false;
    vehicleConnected = false;
    pointChanging = false;
    additionalDegree = -1;
    speed = -1;
}

void HostData::serialize(Message::Writer& w) {
    w.StartObject();
    w.String("streaming");
    w.Bool(streaming);
    w.String("vehicleConnected");
    w.Bool(vehicleConnected);
    w.String("pointChanging");
    w.Bool(pointChanging);
    if (speed > -1) {
        w.String("speed");
        w.Double(speed);
    }
    if (additionalDegree > -1) {
        w.String("additionalDegree");
        w.Int(additionalDegree);
    }
    if (gpsPosition.isExists()) {
        w.String("gpsPosition");
        gpsPosition.serialize(w);
    }
    if (gravitationalField.isExists()) {
        w.String("gravitationalField");
        gravitationalField.serialize(w);
    }
    if (magneticField.isExists()) {
        w.String("magneticField");
        magneticField.serialize(w);
    }
    if (previousGpsPosition.isExists()) {
        w.String("previousGpsPosition");
        previousGpsPosition.serialize(w);
    }
    if (previousGravitationalField.isExists()) {
        w.String("previousGravitationalField");
        previousGravitationalField.serialize(w);
    }
    if (previousMagneticField.isExists()) {
        w.String("previousMagneticField");
        previousMagneticField.serialize(w);
    }
    BaseData<HostData>::serialize(w);
    w.EndObject();
}

void HostData::deserialize(Message::Document& d) {
    if (d.IsObject()) {
        if (d.HasMember("streaming")) {
            Message::Value& v = d["streaming"];
            if (v.IsBool()) streaming = v.GetBool();
        }
        if (d.HasMember("vehicleConnected")) {
            Message::Value& v = d["vehicleConnected"];
            if (v.IsBool()) vehicleConnected = v.GetBool();
        }
        if (d.HasMember("pointChanging")) {
            Message::Value& v = d["pointChanging"];
            if (v.IsBool()) pointChanging = v.GetBool();
        }
        if (d.HasMember("additionalDegree")) {
            Message::Value& v = d["additionalDegree"];
            if (v.IsInt()) additionalDegree = v.GetInt64();
        }
        if (d.HasMember("speed")) {
            Message::Value& v = d["speed"];
            if (v.IsDouble()) speed = v.GetDouble();
        }
        if (d.HasMember("gpsPosition")) {
            gpsPosition.deserialize(d["gpsPosition"]);
        }
        else {
            gpsPosition.setExists(false);
        }
        if (d.HasMember("gravitationalField")) {
            gravitationalField.deserialize(d["gravitationalField"]);
        }
        else {
            gravitationalField.setExists(false);
        }
        if (d.HasMember("magneticField")) {
            magneticField.deserialize(d["magneticField"]);
        }
        else {
            magneticField.setExists(false);
        }
        if (d.HasMember("previousGpsPosition")) {
            previousGpsPosition.deserialize(d["previousGpsPosition"]);
        }
        else {
            previousGpsPosition.setExists(false);
        }
        if (d.HasMember("previousGravitationalField")) {
            previousGravitationalField.deserialize(d["previousGravitationalField"]);
        }
        else {
            previousGravitationalField.setExists(false);
        }
        if (d.HasMember("previousMagneticField")) {
            previousMagneticField.deserialize(d["previousMagneticField"]);
        }
        else {
            previousMagneticField.setExists(false);
        }
        BaseData<HostData>::deserialize(d);
    }
}

double HostData::getSpeed() {
    return speed;
}
        
int HostData::getAdditionalDegree() {
    return additionalDegree;
}

bool HostData::isStreaming() {
    return streaming;
}

bool HostData::isVehicleConnected() {
    return vehicleConnected;
}

bool HostData::isPointChanging() {
    return pointChanging;
}

Point3D HostData::getGpsPosition() {
    return gpsPosition;
}

Point3D HostData::getGravitationalField() {
    return gravitationalField;
}

Point3D HostData::getMagneticField() {
    return magneticField;
}

Point3D HostData::getPreviousGpsPosition() {
    return previousGpsPosition;
}

Point3D HostData::getPreviousGravitationalField() {
    return previousGravitationalField;
}

Point3D HostData::getPreviousMagneticField() {
    return previousMagneticField;
}

void HostData::setSpeed(double d) {
    speed = d;
}

void HostData::setAdditionalDegree(int i) {
    additionalDegree = i;
}

void HostData::setStreaming(bool b) {
    streaming = b;
}

void HostData::setVehicleConnected(bool b) {
    vehicleConnected = b;
}

void HostData::setPointChanging(bool b) {
    pointChanging = b;
}

void HostData::setGpsPosition(Point3D p) {
    gpsPosition = p;
}

void HostData::setGravitationalField(Point3D p) {
    gravitationalField = p;
}

void HostData::setMagneticField(Point3D p) {
    magneticField = p;
}

void HostData::setPreviousGpsPosition(Point3D p) {
    previousGpsPosition = p;
}

void HostData::setPreviousGravitationalField(Point3D p) {
    previousGravitationalField = p;
}

void HostData::setPreviousMagneticField(Point3D p) {
    previousMagneticField = p;
}

void HostData::update(Message* msg) {
    BatteryPartialHostData* bapd = dynamic_cast<BatteryPartialHostData*>(msg);
    BooleanPartialHostData* bopd = dynamic_cast<BooleanPartialHostData*>(msg);
    ControlPartialHostData* copd = dynamic_cast<ControlPartialHostData*>(msg);
    PointPartialHostData* popd = dynamic_cast<PointPartialHostData*>(msg);
    SpeedPartialHostData* sppd = dynamic_cast<SpeedPartialHostData*>(msg);
    if (bapd) bapd->apply(this);
    if (bopd) bopd->apply(this);
    if (copd) copd->apply(this);
    if (popd) popd->apply(this);
    if (sppd) sppd->apply(this);
}

void HostData::update(HostData* data) {
    setSpeed(data->getSpeed());
    setStreaming(data->isStreaming());
    setVehicleConnected(data->isVehicleConnected());
    setPointChanging(data->isPointChanging());
    setAdditionalDegree(data->getAdditionalDegree());
    setGpsPosition(data->getGpsPosition());
    setGravitationalField(data->getGravitationalField());
    setMagneticField(data->getMagneticField());
    setPreviousGpsPosition(data->getPreviousGpsPosition());
    setPreviousGravitationalField(data->getPreviousGravitationalField());
    setPreviousMagneticField(data->getPreviousMagneticField());
    BaseData<HostData>::update(data);
}
